### PR TITLE
snap-exec: support nested environment variables in environment:

### PR DESCRIPTION
**Pull request by mvo5**
This allows to use nested env vars in the `environment:` block of a snap. It supports both the pre-existing environment (PATH, PS1, SNAP_*) and from within the environment. E.g.:
```
environment:
  FOO: 1
  BAR: $FOO/2
  BAZ: $PATH:/something
```
will work. 

The current search order is: 
1. app specific environment
2. snap specific environment
3. os environment

### CODE REVIEW

**Comment by pedronis** [comment]
what happens with multi-level ones?

FOO: 1
BAR: $FOO/2
BAZ: $BAR/3

circular ones?

File: osutil/env.go [hunk 1]
```diff
@@ -50,3 +53,56 @@ func GetenvBool(key string, dflt ...bool) bool {
...
+	for _, s := range env {
+		l := strings.SplitN(s, "=", 2)
+		if len(l) < 2 {
+			continue
+		}
+		envMap[l[0]] = l[1]
+	}
+
+	// this always terminates, in each iteration of the loop we
+	// eliminate at least one value with a "$var"
```
Comment (niemeyer) [inline osutil/env.go]: Note that at the moment something defined in the outer-most level may refer to a variable that is in the innermost scope. I started with that same logic in spread and regretted for a few reasons. Ended up going down to a more sequential and shell-like logic where what comes first may be used later, even in the local scope. This forces some basic organization, which ends up also making it more readable for people to parse.

[Here](https://github.com/snapcore/spread/blob/master/spread/project.go#L186) is the piece that reads the yaml preserving the map ordering, and ensuring some basic consistency (no double keys, for example).

Comment (mvo5) [inline osutil/env.go]: Thank you! I updated the code and it makes heavy use of the spread environment code now. I called it `strutil.OrderedMap` but I'm happy to adjust as needed. There are some tiny derivations from the spread code but happy to fix/update if I missed something (or if you disagree with the derivations).

File: strutil/map_test.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	c.Check(om.Get("K1"), Equals, "bar")
+	c.Check(om.Get("K2"), Equals, "baz")
+	c.Check(om.Get("K0"), Equals, "foo")
+
+	om.Del("K2")
+	c.Assert(om.Keys(), DeepEquals, []string{"K1", "K0"})
+
+	om.Set("K9", "foobar")
+	c.Assert(om.Keys(), DeepEquals, []string{"K1", "K0", "K9"})
+	c.Check(om.Get("K9"), Equals, "foobar")
```
Comment (niemeyer) [inline strutil/map_test.go]: Can we please get an extra check ensuring we can replace a key in the middle of the keys and still get the proper semantics on keys and retrieving the modified value?  So far we don't have any tests ensuring a Set with changes works well, I think.

**Pull request merged**